File: accounts-daemon/dbus/mailmessage.cpp
//
// Created by victor on 10/11/22.
//

#include "mailmessage.h"
#include "src/mimehtml.h"
#include "src/mimemessage.h"
#include "utils.h"
#include <QDBusMetaType>
#include <QFutureWatcher>

struct MailMessagePrivate {
        static quint64 nextId;
        QString path;
//        QSharedPointer<MimeMessage> message;

        QString to;
        QString subject;
        QString htmlContent;
        QString textContent;

        QString from;
        QString fromAddress;
};

quint64 MailMessagePrivate::nextId = 0;

MailMessage::MailMessage(const QString& to, QObject* parent) :
    QObject(parent) {
    qDBusRegisterMetaType<QPair<QString, QString>>();

    d = new MailMessagePrivate();
    d->path = QStringLiteral("/com/vicr123/accounts/mail/Message%1").arg(d->nextId++);
    d->from = qEnvironmentVariable("SMTP_SENDER_NAME");
    d->fromAddress = qEnvironmentVariable("SMTP_SENDER_EMAIL");
    d->to = to;

    Utils::accountsBus().registerObject(d->path, this, QDBusConnection::ExportScriptableContents);
}

MailMessage::~MailMessage() {
    delete d;
}

QDBusObjectPath MailMessage::path() {
    return QDBusObjectPath(d->path);
}

void MailMessage::setSubject(const QString& subject) {
    d->subject = subject;
}

QString MailMessage::subject() {
    return d->subject;
}

[[maybe_unused]] void MailMessage::Send(const QDBusMessage& message) {
    message.setDelayedReply(true);

    auto mailMessage = QSharedPointer<MimeMessage>(new MimeMessage());
    mailMessage->addTo(new EmailAddress(d->to));
    mailMessage->setSender(new EmailAddress(d->fromAddress, d->from));
    mailMessage->setSubject(d->subject);

    auto* htmlPart = new MimeHtml();
    htmlPart->setHtml(d->htmlContent);

    auto textPart = new MimeText();
    textPart->setText(d->textContent);

    mailMessage->addPart(htmlPart);
    mailMessage->addPart(textPart);

    auto watcher = new QFutureWatcher<void>(this);
    watcher->setFuture(Utils::sendMailMessage(mailMessage));
    connect(watcher, &QFutureWatcher<void>::finished, this, [message, this, watcher] {
        if (watcher->isCanceled()) return;

        auto reply = message.createReply();
        Utils::accountsBus().send(reply);

        this->deleteLater();
    });
    connect(watcher, &QFutureWatcher<void>::canceled, this, [message] {
        Utils::sendDbusError(Utils::EmailError, message);
    });
}

[[maybe_unused]] void MailMessage::Discard(const QDBusMessage& message) {
    this->deleteLater();
}

void MailMessage::setFrom(const QPair<QString, QString>& from) {
    d->fromAddress = from.first;
    d->from = from.second;
}

QPair<QString, QString> MailMessage::from() {
    return {
        d->fromAddress,
            d->from
    };
}

QString MailMessage::htmlContent() {
    return d->htmlContent;
}

void MailMessage::setHtmlContent(QString htmlContent) {
    d->htmlContent = htmlContent;
}

QString MailMessage::textContent() {
    return d->textContent;
}

void MailMessage::setTextContent(QString textContent) {
    d->textContent = textContent;
}

File: src/mail_message.rs
use crate::error::Error;
use crate::send_mail_message;
use mail_send::mail_builder::MessageBuilder;
use zbus::{interface, Connection};
use zvariant::OwnedObjectPath;

pub struct MailMessage {
    to: String,
    subject: String,
    from_address: String,
    from: String,
    text_content: String,
    html_content: String,

    path: OwnedObjectPath,
}

impl MailMessage {
    pub fn new(path: OwnedObjectPath, to: &str) -> Self {
        Self {
            to: to.to_string(),
            html_content: Default::default(),
            text_content: Default::default(),
            subject: Default::default(),
            from: Default::default(),
            from_address: Default::default(),
            path,
        }
    }
}

#[interface(name = "com.vicr123.accounts.MailMessage")]
impl MailMessage {
    #[zbus(property)]
    pub fn subject(&self) -> String {
        self.subject.clone()
    }

    #[zbus(property)]
    pub fn set_subject(&mut self, subject: String) {
        self.subject = subject;
    }

    #[zbus(property)]
    pub fn from(&self) -> (String, String) {
        (self.from_address.clone(), self.from.clone())
    }

    #[zbus(property)]
    pub fn set_from(&mut self, from: (String, String)) {
        let (from_address, from) = from;
        self.from_address = from_address;
        self.from = from;
    }

    #[zbus(property)]
    pub fn html_content(&self) -> String {
        self.html_content.clone()
    }

    #[zbus(property)]
    pub fn set_html_content(&mut self, html_content: String) {
        self.html_content = html_content;
    }

    #[zbus(property)]
    pub fn text_content(&self) -> String {
        self.text_content.clone()
    }

    #[zbus(property)]
    pub fn set_text_content(&mut self, text_content: String) {
        self.text_content = text_content;
    }

    pub async fn send(&self, #[zbus(connection)] connection: &Connection) -> Result<(), Error> {
        let message = MessageBuilder::new()
            .from((self.from.clone(), self.from_address.clone()))
            .to(self.to.clone())
            .subject(self.subject.clone())
            .html_body(self.html_content.clone())
            .text_body(self.text_content.clone());

        send_mail_message(message).await?;

        let path = self.path.clone();
        let connection = connection.clone();

        tokio::spawn(async move {
            let _ = connection
                .object_server()
                .remove::<Self, _>(&path)
                .await;
        });
        
        Ok(())
    }

    pub async fn discard(&self, #[zbus(connection)] connection: &Connection) {
        let path = self.path.clone();
        let connection = connection.clone();

        tokio::spawn(async move {
            let _ = connection
                .object_server()
                .remove::<Self, _>(&path)
                .await;
        });
    }
}
